editors/code/src/installation/language_server.ts

```diff
@@ -0,0 +1,119 @@
+import { unwrapNotNil } from "ts-not-nil";
+import { spawnSync } from "child_process";
+import * as vscode from "vscode";
+import * as path from "path";
+import { strict as assert } from "assert";
+import { promises as fs } from "fs";
+
+import { BinarySource } from "./interfaces";
+import { fetchLatestArtifactMetadata } from "./fetch_latest_artifact_metadata";
+import { downloadFile } from "./download_file";
+
+export async function downloadLatestLanguageServer(
+    {file: artifactFileName, dir: installationDir, repo}: BinarySource.GithubRelease
+) {
+    const binaryMetadata = await fetchLatestArtifactMetadata(repo, artifactFileName);
+
+    const {
+        releaseName,
+        downloadUrl
+    } = unwrapNotNil(binaryMetadata, `Latest GitHub release lacks "${artifactFileName}" file`);
+
+    await fs.mkdir(installationDir).catch(err => assert.strictEqual(
+        err?.code,
+        "EEXIST",
+        `Couldn't create directory "${installationDir}" to download `+
+        `language server binary: ${err.message}`
+    ));
+
+    const installationPath = path.join(installationDir, artifactFileName);
+
+    await vscode.window.withProgress(
+        {
+            location: vscode.ProgressLocation.Notification,
+            cancellable: false, // FIXME: add support for canceling download?
+            title: `Downloading language server ${releaseName}`
+        },
+        async (progress, _cancellationToken) => {
+            let lastPrecentage = 0;
+            await downloadFile(downloadUrl, installationPath, (readBytes, totalBytes) => {
+                const newPercentage = (readBytes / totalBytes) * 100;
+                progress.report({
+                    message: newPercentage.toFixed(0) + "%",
+                    increment: newPercentage - lastPrecentage
+                });
+
+                lastPrecentage = newPercentage;
+            });
+        }
+    );
+
+    await fs.chmod(installationPath, 755); // Set (rwx, r_x, r_x) permissions
+}
+export async function ensureLanguageServerBinary(
+    langServerSource: null | BinarySource
+): Promise<null | string> {
```

Comment:
Nice idempotent interface here, :+1:

Reply:
Thanks, though this is not strictly idempotent, since it uses global state :sweat_smile:

Reply:
It is idempotent, because any calls after the first with the same argument don't affect anything.